### PR TITLE
[expo-auth-session] - Allow server side PKCE with secret verifier

### PR DESCRIPTION
**Pull request by stevesouth**
# Why

In authorisation flows where the oath 'client' is a server and will ultimately do the code for token exchange, it is more secure for the code verifier and code challenge to be generated server side, and for the verifier to never be sent to the client.

Currently within AuthRequest it is not possible to only use a pre-generated code challenge, as the code I am changing here checks to see if codeVerifier is null, and if so generates both a new code verifier and a code challenge.

# How

Changing this check to codeChallenge means it is possible to skip a new pair being generated if my auth sever has already sent me a code challenge to use. It can later use the matching verifier which has never been sent out to the client.

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-auth-session/src/AuthRequest.ts [hunk 1]
```diff
@@ -263,7 +263,7 @@ export class AuthRequest implements Omit<AuthRequestConfig, 'state'> {
   }
 
   private async ensureCodeIsSetupAsync(): Promise<void> {
-    if (this.codeVerifier) {
+    if (this.codeChallenge) {
```
Comment (EvanBacon) [inline packages/expo-auth-session/src/AuthRequest.ts:266]: Add a detailed comment here explaining the reasoning. In theory there shouldn't be anything crucial breaking, but it's hard to know for sure given the public scope of `codeVerifier`.

Comment (stevesouth) [reply packages/expo-auth-session/src/AuthRequest.ts:266]: Agreed. I'll try and get the local environment all set up and work through the flows in more detail. I'm new to this code though and oauth in general, so any feedback is welcome.

Essentially what this code currently seems to do is either generate both a verifier and a challenge, or allow the developer to pass in both their own verifier and challenge. 

Technically the developer could only pass in a verifier to stop the auto generation, however that would then not work as there would be a missing challenge on the code request.

I think changing this check to codeChallenge makes more sense as then the options are:

1) Pass in neither the verifier or the challenge and they are both generated for you. This means the PKCE flow is all done within the mobile client.

2) Pass in both a verifier and a challenge if the developer already has a pair generated locally, and still the PKCE flow is done with the mobile client.

3) Pass in only a challenge as the developer has received a challenge for a server endpoint, but the verifier is only known to the server, which will do the code for token exchange later server side.

Hopefully this makes sense. The other reason for this change is i lost a few hours passing in only a code challenge and kept getting 'Invalid code verifier' back from google as the library was auto generating a new pair. I can probably add some additional errors or warnings around this too.

Comment (stevesouth) [reply packages/expo-auth-session/src/AuthRequest.ts:267]: Another gap i've spotted is that AuthRequest doesn't even attempt to use the codeChallenge on AuthRequestConfig. I have now fixed this in the PR too and verified (with the google provider) that the flow now works if you use your own codeChallenge.

**Comment by stevesouth** [comment]
Given the redirect url needs to match for the code request and the code to token exchange I'm now unsure this change is useful, at least not for the authentication flow I was attempting.

**Comment by boredland** [comment]
> Given the redirect url needs to match for the code request and the code to token exchange I'm now unsure this change is useful, at least not for the authentication flow I was attempting.

Could you elaborate a bit more on that? I'd say this could still be valid and your demo-app as I see it still relies on it:
https://github.com/stevesouth/next-auth-react-native-expo/blob/main/demo-app/LoginGoogle.js#L25

**Pull request closed without merging**